llvm/lib/CASObjectFormats/SimpleV1.cpp

```diff
@@ -0,0 +1,1217 @@
+
+  // Nothing remains, no edges.
+  if (!BlockInfo->Remaining)
+    return Error::success();
+
+  if (!NestEdgesInBlock) {
+    auto Edge = LGB.getNode<EdgeListRef>(BlockInfo->BlockIdx++);
+    if (!Edge)
+      return Edge.takeError();
+
+    if (auto Err = Edge->materialize(LGB, *BlockInfo->Block))
+      return Err;
+
+    return Error::success();
+  }
+
+  auto Remaining = getData().take_back(BlockInfo->Remaining);
+  unsigned EdgeSize;
+  auto E = encoding::consumeVBR8(Remaining, EdgeSize);
+  if (E)
+    return E;
+
+  for (unsigned I = 0; I < EdgeSize; ++I) {
+    unsigned SymbolIdx;
+    E = encoding::consumeVBR8(Remaining, SymbolIdx);
+    if (E)
+      return E;
+
+    auto Symbol = LGB.getSymbol(SymbolIdx);
+    if (!Symbol)
+      return Symbol.takeError();
+
+    unsigned char Bits, Kind;
+    unsigned Offset, Addend, Relocation = 0;
+    E = encoding::consumeVBR8(Remaining, Bits);
+    if (!E)
+      E = encoding::consumeVBR8(Remaining, Kind);
+    if (!E)
+      E = encoding::consumeVBR8(Remaining, Offset);
+    if (!E)
+      E = encoding::consumeVBR8(Remaining, Addend);
+    if (E)
+      return E;
+
+    bool IsRelocation = Bits & (1U << 1);
+    if (IsRelocation)
+      E = encoding::consumeVBR8(Remaining, Relocation);
+    if (E)
+      return E;
+
+    BlockInfo->Block->addEdge((jitlink::Edge::Kind)Kind, Offset, **Symbol,
+                              Addend);
+  }
+
+  return Error::success();
+}
+
+Expected<BlockRef> BlockRef::get(Expected<ObjectFormatNodeRef> Ref) {
+  auto Specific = SpecificRefT::getSpecific(std::move(Ref));
+  if (!Specific)
+    return Specific.takeError();
+
+  return BlockRef(*Specific);
+}
+
+Expected<SymbolRef> SymbolRef::create(CompileUnitBuilder &CUB,
+                                      const jitlink::Symbol &S) {
+  Expected<Builder> B = Builder::startNode(CUB.Schema, KindString);
+  if (!B)
+    return B.takeError();
+
+  auto Name = CUB.getOrCreateName(S.getName());
+  if (!Name)
+    return Name.takeError();
+
+  // Encode attributes. FIXME: Not optimal encoding.
+  encoding::writeVBR8(S.getSize(), B->Data);
+  encoding::writeVBR8(S.getOffset(), B->Data);
+  unsigned Bits = 0;
+  Bits |= (unsigned)S.getScope();        // 2 bits
+  Bits |= (unsigned)S.getLinkage() << 2; // 1 bit
+  Bits |= (unsigned)S.isDefined() << 3;  // 1 bit
+  Bits |= (unsigned)S.isLive() << 4;     // 1 bit
+  Bits |= (unsigned)S.isCallable() << 5; // 1 bit
+  Bits |= (unsigned)S.isAutoHide() << 6; // 1 bit
+  encoding::writeVBR8(Bits, B->Data);
+
+  if (S.isDefined()) {
+    auto BlockIndex = CUB.getBlockIndex(S.getBlock());
+    if (!BlockIndex)
+      return BlockIndex.takeError();
+    encoding::writeVBR8(*BlockIndex, B->Data);
+  }
+
+  return get(B->build());
+}
+
+Error SymbolRef::materialize(LinkGraphBuilder &LGB, unsigned Idx) const {
+  StringRef Remaining = getData();
+  unsigned Size, Offset, Bits;
+
+  auto Name = LGB.nextNode<NameRef>();
+  if (!Name)
+    return Name.takeError();
```

Comment:
Ah, that's subtle.

Since SymbolRefs won't be separate CAS objects, you can just use a 4B index in the list of CAS objects.

(Would be nice to have a histogram of names by size. There may be a lot that are under 20B or so, in which case it'd be nice to have a split string table:
- Small names in a string pool, use a positive index
- Big names as separate CAS objects
... but for now splitting them all out seems fine...)

Reply:
I redo the function signature to return `Error` and rename it to `createAndReference*` so it doesn't look like we drop the ref since it is useless.